### PR TITLE
Launch Dataflow templates from GAE Standard

### PR DESCRIPTION
**Pull request by justinwhite**
Adds a trivial example of launching a dataflow template via REST.

Note that this commit adds a new directory, that is a copy of appengine-standard-java8/helloworld with a minor delta to launch a dataflow job.

Below are the complete steps required to make this work, but the general steps are outlined in README.md.

```
export ORGANIZATION=
export BILLING_ACCOUNT=
git clone https://github.com/GoogleCloudPlatform/getting-started-java.git
export PROJECT=dataflow-$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 10 | head -n 1)
export BUCKET=$PROJECT-template
gcloud projects create $PROJECT --set-as-default --organization=$ORGANIZATION
gcloud beta billing projects link $PROJECT --billing-account=$BILLING_ACCOUNT
gcloud services enable dataflow.googleapis.com
gcloud app create --region=us-central
gsutil mb gs://$BUCKET
cd getting-started-java/appengine-standard-java8/helloworld-gae-launch-dataflow
sed -i "s/YOUR_PROJECT_NAME/$PROJECT/" src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java
sed -i "s/YOUR_BUCKET_NAME/$BUCKET/" src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java
mvn appengine:deploy
gcloud app browse
gcloud projects delete $PROJECT # cleanup
```

### CODE REVIEW

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java]: should be just 1 *

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/webapp/index.jsp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/webapp/index.jsp]: Is this file needed?

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  @After public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGetWritesResponse() throws Exception {
+    servletUnderTest.doGet(mockRequest, mockResponse);
+  }
+
+  @Test
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java]: Do we need this?

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Unit tests for {@link HelloAppEngine}.
+ */
+// [START example]
+@RunWith(JUnit4.class)
+public class HelloAppEngineTest {
+  private static final String FAKE_URL = "fake.fk/hello";
+  // Set up a helper so that the ApiProxy returns a valid environment for local testing.
+  private final LocalServiceTestHelper helper = new LocalServiceTestHelper();
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/test/java/com/example/appengine/java8/HelloAppEngineTest.java]: Is there any specific testing?

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      reader.close();
+
+    } else {
+      StringWriter w = new StringWriter();
+      IOUtils.copy(conn.getErrorStream(), w, "UTF-8");
+      response.getWriter().println(w.toString());
+    }
+  }
+
+  public static String getInfo() {
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: Is this helpful?

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      e.printStackTrace();
+    }
+
+    URL url = new URL("https://dataflow.googleapis.com/v1b3/projects/"
+        + project + "/templates:launch?gcs_path="
+        + "gs://dataflow-templates/latest/Datastore_to_GCS_Text");
+    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+    conn.setDoOutput(true);
+    conn.setRequestMethod("POST");
+    conn.setRequestProperty("Authorization", "Bearer " + accessToken.getAccessToken());
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: A more current way  (for Java8) is to get this from the Metadata server instead of using the API.

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+public class HelloAppEngine extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+
+    String project = "YOUR_PROJECT_NAME";
+    String bucket = "gs://YOUR_BUCKET_NAME";
+
+    ArrayList<String> scopes = new ArrayList<String>();
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: Using the Metadata server is a better practice.

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.io.IOUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+// With @WebServlet annotation the webapp/WEB-INF/web.xml is no longer required.
+@WebServlet(name = "HelloAppEngine", value = "/hello")
+public class HelloAppEngine extends HttpServlet {
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: Should rename for dataflow

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
+/**
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: Should be a new name.
Only a single *

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/settings.gradle [hunk 10]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/settings.gradle]: Probably don't need the gradle version.

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+  <build>
+    <!-- for hot reload of the web application-->
+    <outputDirectory>${project.build.directory}/${project.build.finalName}/WEB-INF/classes</outputDirectory>
+    <plugins>
+      <!-- [START cloudplugin] -->
+      <plugin>
+        <groupId>com.google.cloud.tools</groupId>
+        <artifactId>appengine-maven-plugin</artifactId>
+        <version>1.3.1</version>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml]: 1.3.2 ??

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+-->
+<!-- [START pom] -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+  <!-- REQUIRED : your application must be packed as a .war to deploy on App Engine Standard -->
+  <packaging>war</packaging>
+
+  <groupId>com.example.appengine-j8</groupId>
+  <artifactId>helloworld</artifactId>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml]: new name

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+  <version>1.0-SNAPSHOT</version>
+
+  <!-- [START_EXCLUDE] -->
+  <parent>                              <!-- Get parameters and allow bulk testing -->
+    <artifactId>getting-started-java</artifactId>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/pom.xml]: Parent pom should be:
```
  <parent>
    <groupId>com.google.cloud.samples</groupId>
    <artifactId>shared-configuration</artifactId>
    <version>1.0.8</version>
  </parent>
```

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradlew.bat [hunk 14]
```diff
@@ -0,0 +1,84 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradlew.bat]: don't need file

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradlew [hunk 15]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradlew]: don't include file

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradle/wrapper/gradle-wrapper.properties [hunk 16]
```diff
@@ -0,0 +1,6 @@
+#Tue Jun 13 16:53:48 PDT 2017
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/gradle/wrapper/gradle-wrapper.properties]: don't include file

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/build.gradle [hunk 17]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2017 Google Inc.
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/build.gradle]: don't include file

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/README.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+If you do not have gradle installed, you can deploy using `./gradlew appengineDeploy`.
+
+To use vist:  https://YOUR-PROJECT-ID.appspot.com
+
+## Testing
+
+    mvn verify
+
+ or
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/README.md]: skip gradle

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/README.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+## Gradle
+### Running locally
+
+    gradle appengineRun
+
+If you do not have gradle installed, you can run using `./gradlew appengineRun`.
+
+To use vist: http://localhost:8080/
+
+### Deploying
```
Comment (lesv) [inline appengine-standard-java8/helloworld-launch-dataflow-from-gae/README.md]: skip gradle

**Comment by lesv** [comment]
One more thing, to hook it into our testing, just add the name to the parent pom.xml (in the folder above, so that our automated builds pick it up.

File: appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java (hunk 5 above)
Comment (justinwhite) [reply appengine-standard-java8/helloworld-launch-dataflow-from-gae/src/main/java/com/example/appengine/java8/HelloAppEngine.java]: My goal was to minimize the delta between this example and the original HelloAppEngine sample. Removing this meant ripping a chunk of the tests out as well. I have no problem removing both for simplicity's sake now this this example is being distanced from the original.

File: appengine-standard-java8/helloworld-gae-launch-dataflow/README.md [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+    ```
+    gcloud services enable dataflow.googleapis.com
+    ```
+
+1. Set the project and bucket names (required for the following steps). These
+   commands add generic names; feel free to adjust appropriately.
+
+    ```
+    export PROJECT=dataflow-$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 10 | head -n 1)
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: This might not be the best thing for users, though it might help us w/ testing.

File: appengine-standard-java8/helloworld-gae-launch-dataflow/README.md [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+1. Create the GCS bucket that dataflow will use
+
+    ```
+    gsutil mb gs://$BUCKET
+    ```
+
+1. Update the source to use your project and bucket names in the dataflow job.
+
+    ```
+    sed -i "s/YOUR_PROJECT_NAME/$PROJECT/" src/main/java/com/example/appengine/java8/HelloAppEngine.java
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: this probably changed

File: appengine-standard-java8/helloworld-gae-launch-dataflow/README.md [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+To use vist: http://localhost:8080/
+
+### Deploying
+
+    mvn appengine:deploy
+
+To use vist:  https://YOUR-PROJECT-ID.appspot.com
+
+## Gradle
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: Let's remove Gradle

File: appengine-standard-java8/helloworld-gae-launch-dataflow/pom.xml [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+    <dependency>
+      <groupId>com.google.appengine</groupId>
+      <artifactId>appengine-api-1.0-sdk</artifactId>
+      <version>1.9.60</version>
+    </dependency>
+    <!-- [START servlet] -->
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <version>4.0.0</version>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/pom.xml]: This probably should be 3.1.0

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+        .put("bypassTempDirValidation", false);
+      jsonObj = new JSONObject()
+        .put("jobName", "template-" + UUID.randomUUID().toString())
+        .put("parameters", parameters)
+        .put("environment", environment);
+    } catch (JSONException e) {
+      e.printStackTrace();
+    }
+
+    URL url = new URL("https://dataflow.googleapis.com/v1b3/projects/"
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java]: This might be better w/ `String.format()`

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+      reader.close();
+
+    } else {
+      StringWriter w = new StringWriter();
+      IOUtils.copy(conn.getErrorStream(), w, "UTF-8");
+      response.getWriter().println(w.toString());
+    }
+  }
+
+  public static String getInfo() {
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java]: Is this required?

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java [hunk 26]
```diff
@@ -0,0 +1,112 @@
+/**
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java]: 2nd `*` not required.  (ie it should be `/*` not `/**`

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/webapp/index.jsp [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+<!-- [END_EXCLUDE] -->
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<%@ page import="com.example.appengine.java8.HelloAppEngine" %>
+<html>
+<head>
+  <link href='//fonts.googleapis.com/css?family=Marmelad' rel='stylesheet' type='text/css'>
+  <title>Hello App Engine Standard Java 8</title>
+</head>
+<body>
+    <h1>Hello App Engine -- Java 8!</h1>
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/webapp/index.jsp]: Please change to mention the sample you are using.

Comment (lesv) [reply appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/webapp/index.jsp]: Or remove the file

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.io.IOUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+// With @WebServlet annotation the webapp/WEB-INF/web.xml is no longer required.
+@WebServlet(name = "LaunchDataflowTemplate", value = "/launch")
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java]: `/launchdf`  ??

File: appengine-standard-java8/helloworld-gae-launch-dataflow/src/test/java/com/example/appengine/java8/LaunchDataflowTemplateTest.java [hunk 29]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/src/test/java/com/example/appengine/java8/LaunchDataflowTemplateTest.java]: Copyrights should only have `/*`

File: appengine-standard-java8/helloworld-gae-launch-dataflow/README.md [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+
+1. Update the source to use your project and bucket names in the dataflow job.
+
+    ```
+    sed -i "s/YOUR_PROJECT_NAME/$PROJECT/" src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java
+    sed -i "s/YOUR_BUCKET_NAME/$BUCKET/" src/main/java/com/example/appengine/java8/LaunchDataflowTemplate.java
+    ```
+
+## Maven
+### Running locally
```
Comment (lesv) [inline appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: Does it actually run locally?

Comment (justinwhite) [reply appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: Yes. Looks like something with gcloud 187 is breaking the deployment though. Working on the issue now.

Comment (justinwhite) [reply appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: Looks like downgrading to 186 fixed the issue.. going to try to file a bug
```
sudo gcloud components update --version 186.0.0
```

Comment (justinwhite) [reply appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: here's the issue: https://stackoverflow.com/questions/48565534/failed-to-generate-quickstart-xml-while-deploying-on-app-engine-standard

Comment (lesv) [reply appengine-standard-java8/helloworld-gae-launch-dataflow/README.md]: Argh - ok, mention it in the README.
